### PR TITLE
[Distributed] Fix `_executeDistributedFunction` resume and result buffer

### PR DESCRIPTION
**Pull request by xedin**
* Setup correct context so that `swift_distributed_execute_target` resume back to the correct fame.

Since `swift_distributed_execute_target` is not a regular async function, it needs to do some extra
bookkeeping before accessor could be called.

* Drop optionality from result buffer, works fine with `Void` and makes it simpler to implement the accessor.

* Add some more tests to make sure that different result types are handled correctly.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

File: stdlib/public/Distributed/DistributedActorSystem.swift [hunk 1]
```diff
@@ -277,7 +274,7 @@ func _executeDistributedTarget(
   on actor: AnyObject, // DistributedActor
   _ targetName: UnsafePointer<UInt8>, _ targetNameLength: UInt,
   argumentBuffer: Builtin.RawPointer, // HeterogeneousBuffer of arguments
-  resultBuffer: Builtin.RawPointer?
+  resultBuffer: Builtin.RawPointer
```
Comment (ktoso) [inline stdlib/public/Distributed/DistributedActorSystem.swift:277]: Gotcha, noted 👍

File: stdlib/public/Concurrency/Actor.cpp [hunk 2]
```diff
@@ -1741,7 +1750,17 @@ void ::swift_distributed_execute_target(
...
+  // to pass "caller" context and resume function as extra parameters due to
+  // how they are declared in C. But this particular function behaves exactly
+  // like a regular `async throws`, which means that we need to initialize
+  // intermediate `callContext` using parent `callerContext`. A better fix for
+  // this situation would be to adjust IRGen and handle function like this
+  // like regular `async` functions even though they are classified as special.
+  callContext->Parent = callerContext;
+  callContext->ResumeParent = resumeFunc;
+
+  calleeContext->Parent = callContext;
```
Comment (ktoso) [inline stdlib/public/Concurrency/Actor.cpp:1763]: That's tricky... thanks a lot for digging through!

**Review by ktoso — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please test macOS platform

**Comment by ktoso** [comment]
Looking good, let's merge 👍 

I'll send a mini PR for the parameters :)

**Pull request merged**